test/performance/odp_dma_perf.c

```diff
@@ -0,0 +1,876 @@
+
+static int parse_options(int argc, char **argv, test_config_t *config)
+{
+	int opt, long_index;
+
+	static const struct option longopts[] = {
+		{ "trs_type", required_argument, NULL, 't' },
+		{ "trs_grn", required_argument, NULL, 'g' },
+		{ "num_in_seg", required_argument, NULL, 'i' },
+		{ "in_seg_size", required_argument, NULL, 's' },
+		{ "in_seg_type", required_argument, NULL, 'T' },
+		{ "compl_modes", required_argument, NULL, 'm' },
+		{ "num_rounds", required_argument, NULL, 'r' },
+		{ "help", no_argument, NULL, 'h' },
+		{ NULL, 0, NULL, 0 }
+	};
+
+	static const char *shortopts = "t:g:i:s:T:m:r:h";
+
+	set_option_defaults(config);
+
+	while (1) {
+		opt = getopt_long(argc, argv, shortopts, longopts, &long_index);
+
+		if (opt == -1)
+			break;
+
+		switch (opt) {
+		case 't':
+			config->trs_type = atoi(optarg);
+			break;
+		case 'g':
+			config->trs_grn = atoi(optarg);
+			break;
+		case 'i':
+			config->num_in_seg = atoi(optarg);
+			break;
+		case 's':
+			config->seg_size = atoi(optarg);
+			break;
+		case 'T':
+			config->seg_type = atoi(optarg);
+			break;
+		case 'm':
+			parse_completion_modes(config, optarg);
+			break;
+		case 'r':
+			config->num_rounds = atoi(optarg);
+			break;
+		case 'h':
+		default:
+			print_usage();
+			return -1;
+		}
+	}
+
+	if (check_options(config))
+		return -1;
+
+	return 0;
+}
+
+static int check_capabilities(test_config_t *config)
+{
+	odp_dma_capability_t capa;
+	const int is_event = config->compl_modes.compl_mask & ODP_DMA_COMPL_EVENT;
+
+	if (odp_dma_capability(&capa)) {
+		ODPH_ERR("Error querying DMA capabilities.\n");
+		return -1;
+	}
+
+	if (capa.max_sessions == 0U) {
+		ODPH_ERR("DMA not supported.\n");
+		return -1;
+	}
+
+	if (config->trs_type == TRS_TYPE_ASYNC) {
+		if (config->trs_grn == GRN_IND &&
+		    (uint32_t)config->num_in_seg > capa.max_transfers) {
+			ODPH_ERR("Unsupported amount of in-flight DMA transfers.\n");
+			return -1;
+		}
+
+		if ((config->compl_modes.compl_mask & ODP_DMA_COMPL_POLL) &&
+		    (capa.compl_mode_mask & ODP_DMA_COMPL_POLL) == 0U) {
+			ODPH_ERR("Unsupported completion mode, poll.\n");
+			return -1;
+		}
+
+		if (is_event && (capa.compl_mode_mask & ODP_DMA_COMPL_EVENT) == 0U) {
+			ODPH_ERR("Unsupported completion mode, event.\n");
+			return -1;
+		}
+
+		if (is_event && capa.queue_type_sched == 0) {
+			ODPH_ERR("Unsupported queueing type.\n");
+			return -1;
+		}
+	}
+
+	if (config->trs_grn == GRN_ALL) {
+		if ((uint32_t)config->num_in_seg > capa.max_src_segs) {
+			ODPH_ERR("Unsupported amount of source segments.\n");
+			return -1;
+		}
+
+		if (config->num_in_seg + 1U > capa.max_segs) {
+			ODPH_ERR("Unsupported total amount of segments.\n");
+			return -1;
+		}
+	}
+
+	if (config->trs_grn == GRN_IND && capa.max_segs < 2U) {
+		ODPH_ERR("Unsupported total amount of segments.\n");
+		return -1;
+	}
+
+	if (config->num_in_seg * config->seg_size > capa.max_seg_len) {
+		ODPH_ERR("Unsupported total segment size.\n");
+		return -1;
+	}
+
```

Comment:
`odp_dma_capability_t.pool.max_num` could also be checked (see comment below about completion events).

Reply:
Fixed in v4.